mountpoint-s3-crt/src/s3/endpoint_resolver.rs

```diff
@@ -0,0 +1,329 @@
+//! Endpoint resolver helps determine which S3 endpoint to use for a given configuration.
+
+use mountpoint_s3_crt_sys::*;
+use std::{
+    ffi::{OsStr, OsString},
+    os::unix::prelude::OsStrExt,
+    ptr::{self, NonNull},
+};
+
+use crate::{
+    aws_byte_cursor_as_slice,
+    common::{allocator::Allocator, error::Error},
+    CrtError, ToAwsByteCursor,
+};
+
+use super::s3_library_init;
+
+/// [RuleEngine] to resolve endpoint with given [RequestContext] and ruleset
+#[derive(Debug)]
+pub struct RuleEngine {
+    inner: NonNull<aws_endpoints_rule_engine>,
+}
+
+impl RuleEngine {
+    /// Creates a new endpoint [RuleEngine].
+    pub fn new(allocator: &Allocator) -> Result<Self, Error> {
+        s3_library_init(allocator);
+        // SAFETY: `allocator.inner` is a valid aws_allocator and and we check the return is non-null.
+        // SAFETY: aws_s3_endpoint_resolver_new will acquire a reference to keep it alive after this function call, so it's safe to return an owned version to it.
+        let inner = unsafe { aws_s3_endpoint_resolver_new(allocator.inner.as_ptr()).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Resolve the endpoint with given [RequestContext] and ruleset.
+    pub fn resolve(&self, context: RequestContext) -> Result<ResolvedEndpoint, Error> {
+        // SAFETY: `aws_endpoints_rule_engine_resolve` ensures that it returns a `aws_endpoints_resolved_endpoint`
+        // (which is checked for being non-null later) or it will return an error. So, the out_endpoint is valid and non null.
+        // SAFETY: Next, we handle all the arms `aws_endpoints_resolved_endpoint_get_type` separately.
+        unsafe {
+            let mut out_endpoint: *mut aws_endpoints_resolved_endpoint = ptr::null_mut();
+            aws_endpoints_rule_engine_resolve(self.inner.as_ptr(), context.inner.as_ptr(), &mut out_endpoint)
+                .ok_or_last_error()?;
+            match aws_endpoints_resolved_endpoint_get_type(out_endpoint) {
+                aws_endpoints_resolved_endpoint_type::AWS_ENDPOINTS_RESOLVED_ERROR => {
+                    let mut out_error: aws_byte_cursor = Default::default();
+                    let err = Error::from(aws_endpoints_resolved_endpoint_get_error(out_endpoint, &mut out_error));
```

Comment:
I want to get error out of `out_error` which this is not doing.

Reply:
Yeah,`aws_endpoints_resolved_endpoint_get_error` always returns AWS_OP_SUCCESS here. In other words, the return value of `aws_endpoints_resolved_endpoint_get_error` is telling you whether _that operation_ (`get_error`) succeeded or not, which it always will in this branch.

This is why we were talking about making the return type of this method `Result<Result<ResolvedEndpoint, String>, Error>`. The outer `Result` captures "internal" errors, like the JSON not parsing or the rules being invalid. The inner `Result` captures the effect of the actual resolver logic, which sometimes gives an endpoint but sometimes gives an error string.

The `Uri` struct has a bunch of examples of turning `aws_byte_cursor`s into `OsStr`s, which you'll need to tweak slightly because you want to return an owned string here, not a reference.

Reply:
Also, this case is leaking the `aws_endpoints_resolved_endpoint` right now; it needs a `release`.